component/kafka/component.go

```diff
@@ -475,8 +474,9 @@ func (c *consumerHandler) flush(session sarama.ConsumerGroupSession) error {
 			}
 		}
 
-		for _, m := range c.msgBuf {
-			session.MarkMessage(m, "")
+		for _, m := range messages {
+			session.MarkMessage(m.Message(), "")
+			trace.SpanSuccess(m.Span())
```

Comment:
here is what I meant about adding the success `trace` in the component alongside "marking the message".. which does not guarantee commit in case of Sarama. Unless we configure commit sync. WDYT?

Reply:
This will mark all messages in the buffer as processed. or if the proc function fails it will do nothing. Correct?
There is no in between state like marking 5 out of 10 as success.

Reply:
Indeed, marking 5 out of 10 messages will not work with Kafka batching and implies that some messages will be reprocessed. Let me give an example. 

We have messages with the following offsets `1, 2, 3, 4, 5, 6, 7, 8, 9, 10`... 5 messages succeed, say `1, 2, 8, 9, 10`. Then marking message 10 actually **marks the whole batch anyways one you explicitly commit or the ticker in Sarama commits**. So I can't say I necessarily agree with partial marking of messages. If the client want to mark message by message then set batch size to 1 as apparently every message is independent and has its own state.

Reply:
I also agree that we cannot support this correctly but we need to make sure that the users are fully aware because they have to make their systems idempotent in any case!

Reply:
So the idea is that each batch will either commit fully or do not commit at all.
this makes thins a lot simpler and we do not need to have handle commit per message or per batch by the devs.

Reply:
Agree. Of course this is just a bit unorthodox with Sarama, let's discuss the following scenarios which I think will help us reach an agreement. And these scenarios are also related to the sync commit option provided to the consumer.

Normally Sarama as we know commits via ticker https://github.com/Shopify/sarama/blob/master/offset_manager.go#L237-L238

This can be disabled or manipulated through `saramaConfig.Consumer.Offsets.AutoCommit.Enable = false` and `saramaConfig.Consumer.Offsets.AutoCommit.Interval = 1 * time.Second`

 **Scenario 1**
- A developer wants to abide by Sarama committing guarantees, **they only need to `MarkMessage`** and Sarama will commit every x timeunits.
 
 **Scenario 2**
- A developer wants to guarantee committing in a blocking operation which has more guarantees. In this PR there is `WithSyncCommit` but actually we can _**better**_ it to the `Batch` interface and add the `Commit()` method there so a user can choose to `Commit` the batch or not. But here is the catch, should the `Commit()` method do the `MarkMessage` as well if not already marked or should the `MarkMessage` and `trace` be left in the internals of the consumer. Because calling commit without actually marking the message does nothing so we need to be clear that one must `MarkMessage` first. Which also makes the developer wanna loop the messages at the end to mark them. I am more inclined to say the `Batch` commit marks too. But again what happens if autocommit is on? Then calling commit will Mark and commit which means the autocommit is kinda obsolete here and makes the design of our interface not very good.

**Scenario 3**
- A developer never wants to commit, just wants to consume and if the consume restarts, consume everything again. At this point they need to `saramaConfig.Consumer.Offsets.AutoCommit.Enable = false` in the config and never commit a batch. But also at this point `MarkMessage` is unnecessary as you never commit it.

Reply:
An idea is to get around that by introducing 2 methods in the `Batch` interface.
- `Commit()` which is a blocking sync operation: **marks offsets** and **commits them**
- `CommitAsync()` which just **marks offset** and leaves commit to Sarama's ticker

But in this scenario, should `trace.SpanSuccess/Error` be in those methods? What if the developer never calls `Commit()`/`CommitAsync()` do we never call `SpanSuccess`, `SpanError`?

Reply:
Interesting. Maybe we can leave the span management in the component and if the processor function does not return an error the SpanSuccess, if it returns an errot SpanError. What do you think?

Reply:
I think it makes a lot of sense to do that 👍

Reply:
Changed `CommitAsync` to `MarkBatchOffset` https://github.com/beatlabs/patron/pull/342/commits/2f562d73d2ac76366ead7863fd459127ea91da8c

- It didn't make sense to call it `CommitAsync` if it doesn't really commit if `AutoCommit` is false but just marks.

- `Commit` on the other hand, marks then commits.